pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -204,23 +227,62 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 		for _, family := range ipFamily {
 			ips := returnIPsFromSpecificFamily(family, v.IpAddress)
 
-			for _, ip := range ips {
-				klog.V(2).Infof("Adding IP: %s", ip)
+			if addressMatchingEnabled {
+				klog.V(2).Infof("Adding Hostname")
 				v1helper.AddToNodeAddresses(&addrs,
 					v1.NodeAddress{
-						Type:    v1.NodeExternalIP,
-						Address: ip,
-					}, v1.NodeAddress{
-						Type:    v1.NodeInternalIP,
-						Address: ip,
-					}, v1.NodeAddress{
 						Type:    v1.NodeHostName,
 						Address: oVM.Guest.HostName,
 					},
 				)
-
-				found = true
-				break
+				for _, ip := range ips {
+					parsedIP := net.ParseIP(ip)
+					if parsedIP == nil {
+						return err
+					}
+
+					if internalNetworkSubnet != nil && internalNetworkSubnet.Contains(parsedIP) {
+						klog.V(2).Infof("Adding Internal IP: %s", ip)
+						v1helper.AddToNodeAddresses(&addrs,
+							v1.NodeAddress{
+								Type:    v1.NodeInternalIP,
+								Address: ip,
+							},
+						)
+						found = true
+						break
+					}
+
+					if externalNetworkSubnet != nil && externalNetworkSubnet.Contains(parsedIP) {
+						klog.V(2).Infof("Adding External IP: %s", ip)
+						v1helper.AddToNodeAddresses(&addrs,
+							v1.NodeAddress{
+								Type:    v1.NodeExternalIP,
+								Address: ip,
+							},
+						)
+						found = true
+						break
```

Comment:
So I don't think we can actually break here because that means we only store the first address we see

Reply:
We might have to store addresses found in a map and only add the specific address type if it doesn't already exist in the map

Reply:
Could you elaborate on this further?
>that means we only store the first address we see

Isn't that exactly what [we want](https://github.com/kubernetes/cloud-provider-vsphere/pull/271#discussion_r341680191)?

Reply:
We want first address we see of each type, I think at the current state it's only the first address of first type we see, ignoring the other type

Reply:
(because we iterate the same for loop for both types)

Reply:
Oh my god... I've been looking at this `break` statement, and all I've seen was a `continue` statement. Wow.

If this part of the code looks ugly to you, perhaps, I can move it to a function and write a test? Or is it good as it is?

Reply:
Looks great now, thank you!